### ISSUE TITLE
Adding disabled "block" rule results in invalid JSON: unknown field "ruleset"

### ISSUE THREAD

**Issue by waddles** [cloudflare/terraform-provider-cloudflare#2990]
### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

```
❯ terraform -v
Terraform v1.6.5
on darwin_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v4.20.0
```

### Affected resource(s)

* cloudflare_ruleset

### Terraform configuration files

```hcl
resource "cloudflare_ruleset" "custom-firewall-rules" {
  kind    = "zone"
  name    = "default"
  phase   = "http_request_firewall_custom"
  zone_id = "redacted"
  rules {
    action = "block"
    action_parameters {
      ruleset = "current"
    }
    description  = "Block streams"
    enabled      = false
    expression   = "(http.host contains \"streams\")"
    logging {
      enabled = true
    }
  }
}
provider "cloudflare" {
  api_token = "redacted"

  api_client_logging = "true"
}
terraform {
  required_providers {
    cloudflare = {
      source  = "cloudflare/cloudflare"
      version = "4.20.0"
    }
  }
}
```

### Link to debug output

https://gist.github.com/waddles/6a26861ea1d5b8b2b5614a53d957ce9f

### Expected output

When adding with "skip" action, the rule applies cleanly.
```
Plan: 1 to add, 0 to change, 0 to destroy.
cloudflare_ruleset.custom-firewall-rules: Creating...
cloudflare_ruleset.custom-firewall-rules: Creation complete after 2s [id=402d91f22dd8400c8a0b24951fa0e6c5]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

### Actual output

When creating or modifying the rule to use a "block" action, it fails
```
Plan: 0 to add, 1 to change, 0 to destroy.
cloudflare_ruleset.custom-firewall-rules: Modifying... [id=402d91f22dd8400c8a0b24951fa0e6c5]
╷
│ Error: error updating ruleset with ID "402d91f22dd8400c8a0b24951fa0e6c5"
│
│   with cloudflare_ruleset.custom-firewall-rules,
│   on cloudflare-waf.tf line 1, in resource "cloudflare_ruleset" "custom-firewall-rules":
│    1: resource "cloudflare_ruleset" "custom-firewall-rules" {
│
│ invalid JSON: unknown field "ruleset"
```

### Steps to reproduce

```
❯ terraform init
❯ terraform apply -auto-approve
```

**Comment by waddles** [comment]
It does do something, it adds a disabled rule. We work with a number of brands that offer slightly different features but all use the same basic infrastructure. Until now we have been able to use `cloudflare_firewall_rule` resources with a
```
count = var.feature_enabled == true ? 1 : 0
```
type of ternary condition to decide which rules need to be added to the WAF ruleset for each brand.

Moving to a `cloudflare_ruleset` does not allow us to remove some rules from the ruleset in that way, they need to be added to all brands but disabled in the brands that don't need them. Sure, I can keep a separate list of global rules and use a `dynamic {}` block to include only the rules where the conditions are true but that seems unintuitive when the API should just allow me to create a disabled rule.

**Comment by jacobbednarz** [comment]
> Also the error message is completely useless. There is nothing invalid about the JSON and the `ruleset` field is "known" -it's part of the spec.

i agree the error message could be more helpful here however, the validations aren't as straight forward as what you think is on the surface here. some `action`s only have certain `action_parameters` that are valid. for instance, you cannot just any field for a block action that may be valid for a skip action. there are some which overlap but it's not a free for all when it comes to what is supported. this is likely why you are hitting a problem swapping between the actions with the `action_parameters.ruleset` attribute being present.

> Perhaps further to my point, and something that highlights the inconsistency between the Cloudflare Dashboard and what can be achieved with the API:

the dash and API are not 1:1. the API can handle things in their rawest form while the dash wraps some of the functionality to make it easier to visualise for end users. i suspect the way the dash is handling this is building the request from scratch for a PUT whereas the terraform side is taking the existing state (as it is designed to) into account.

cc @zakcutner who may have some more insight here.

**Comment by zakcutner** [comment]
Jacob is correct—the action parameters are completely dependent on what the action is (you can think of them as arguments for the action), but this is a little tricky to express in Terraform. In [the API docs](https://developers.cloudflare.com/api/operations/createZoneRulesetRule), you can see that block rules and log rules take quite different action parameters.

For your specific example, the `ruleset = "current"` action parameter for the skip action causes the rest of the rules in the current ruleset to be skipped. If you would like to both block the request and skip the remaining rules, you can simply use the block action. Blocking the request is a terminating action, so will cause the remainder of the rules to be skipped in any case.

**Comment by waddles** [comment]
@jacobbednarz closing this with no action is pretty lame. There are cases that the terraform provider cannot handle - the whole point of using terraform instead of manually using the Dashboard is so it can be reapplied and so the state of the resources are what is committed in the repo. At the moment, I cannot create the same set of rules with terraform that I can with the Dash, and if I create a rule that is close to what I need then modify it in the Dash, it breaks terraform state and there is no way to recover other than delete the rule outright.

Also you acknowledged the misleading error message but don't want to fix it?

@zakcutner the issue is not about `action_parameters`, it is about creating a rule with `action = "block"` and `enabled = "false"` which according to the API docs you posted above, should be possible. I took your point that `ruleset` is not a valid `action_parameter` when `action = "block"` - another documentation opportunity: I don't know what is valid there - so I removed it and now I get a more sensible error message
```
Terraform will perform the following actions:

  # cloudflare_ruleset.custom-firewall-rules will be updated in-place
  ~ resource "cloudflare_ruleset" "custom-firewall-rules" {
        id      = "402d91f22dd8400c8a0b24951fa0e6c5"
        name    = "default"
        # (3 unchanged attributes hidden)

      ~ rules {
          ~ action       = "skip" -> "block"
          ~ enabled      = false -> true
          ~ id           = "c7c764a36ec641db8c5c1e340662e90d" -> (known after apply)
          ~ last_updated = "2023-12-07 02:07:58.856611 +0000 UTC" -> (known after apply)
          ~ ref          = "c7c764a36ec641db8c5c1e340662e90d" -> (known after apply)
          ~ version      = "1" -> (known after apply)
            # (2 unchanged attributes hidden)

          - action_parameters {
              - ruleset = "current" -> null
            }

          ~ logging {
              ~ enabled = true -> false
            }
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
cloudflare_ruleset.custom-firewall-rules: Modifying... [id=402d91f22dd8400c8a0b24951fa0e6c5]
╷
│ Error: error updating ruleset with ID "402d91f22dd8400c8a0b24951fa0e6c5"
│
│   with cloudflare_ruleset.custom-firewall-rules,
│   on cloudflare-waf.tf line 1, in resource "cloudflare_ruleset" "custom-firewall-rules":
│    1: resource "cloudflare_ruleset" "custom-firewall-rules" {
│
│ logging options only allowed in the skip action (20018)
╵
```

Why can't the provider return a sensible error message like that telling me `ruleset options only allowed in the skip action`?

**Comment by jacobbednarz** [comment]
@waddles the issue is closed as there is no action to take in the provider. the issue here is misunderstanding how the rule actions work coupled with a vague error message. the provider cannot fix either of those as at the end of the day, it is just another API client and we don't wrap or abstract service errors. 

i've raised a ticket internally about the error message improvement which when/if it lands, terraform (along with every other client) will pick up automatically. 

if you're struggling with how to use the ruleset engine within terraform, you're best bet is to do one of the following (depending on what is available to you):

- raise a support ticket asking for clarification on your questions; or
- head over to community.cloudflare.com or the discord for community based support; or
- inspect the HTTP payloads that the dash is making and attempt to map them to terraform attributes (they should be pretty close to 1:1) using the developer docs and registry documentation as reference.

**Comment by bkedryna** [comment]
I just wanted to add my 2 cents. I agree with Wade that error reporting is horrible in the Cloudflare terraform provider. I'm used to working with the aws provider and that is much better. The Cloudflare provider looks like just a simple wrapper around API calls. I just struggled with the same problem and finally figured it out by trial-and-error, but it could have been much easier if error handling was better. Btw, after tweaking something I got an error saying it's a bug in the provider, which supports our claims that this provider needs work:
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to module.waf_rules.cloudflare_ruleset.main[0], provider "provider[\"registry.terraform.io/cloudflare/cloudflare\"].rna" produced an unexpected new value:
│ .rules[5].action_parameters: block count changed from 1 to 0.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
